spec/models/course_spec.rb

```diff
@@ -384,6 +384,230 @@
+      end
+
+      context "course has no vacancies" do
+        let(:site_statuses) do
+          [
+            build(:site_status, :with_no_vacancies),
+          ]
+        end
+
+        it "is not returned" do
+          expect(subject).to be_empty
+        end
+      end
+    end
+
+    describe ".with_study_modes" do
+      let(:course_part_time) { create(:course, study_mode: :part_time) }
+      let(:course_full_time) { create(:course, study_mode: :full_time) }
+      let(:course_both) { create(:course, study_mode: :full_time_or_part_time) }
+
+      subject { described_class.with_study_modes(study_modes) }
+
+      before do
+        course_both
+        course_full_time
+        course_part_time
+      end
+
+      context "full_time" do
+        let(:study_modes) { "full_time" }
+
+        it "returns full time courses" do
+          expect(subject).to contain_exactly(course_both, course_full_time)
+        end
+      end
+
+      context "part_time" do
+        let(:study_modes) { "part_time" }
+
+        it "returns part time courses" do
+          expect(subject).to contain_exactly(course_both, course_part_time)
+        end
+      end
+
+      context "full time and part_time" do
+        let(:study_modes) { %w(full_time part_time) }
+
+        it "returns all" do
+          expect(subject).to contain_exactly(course_both, course_part_time, course_full_time)
+        end
+      end
+    end
+
+    describe ".with_salary" do
+      let(:course_higher_education_programme) do
+        create(:course, program_type: :higher_education_programme)
+      end
+
+      let(:course_school_direct_training_programme) do
+        create(:course, program_type: :school_direct_training_programme)
+      end
+
+      let(:course_school_direct_salaried_training_programme) do
+        create(:course, program_type: :school_direct_salaried_training_programme)
+      end
+
+      let(:course_scitt_programme) { create(:course, program_type: :scitt_programme) }
+      let(:course_pg_teaching_apprenticeship) do
+        create(:course, program_type: :pg_teaching_apprenticeship)
+      end
+
+      subject { described_class.with_salary }
+
+      before do
+        course_higher_education_programme
+        course_school_direct_training_programme
+        course_school_direct_salaried_training_programme
+        course_scitt_programme
+        course_pg_teaching_apprenticeship
+      end
+
+      it "only returns salaried training programme" do
+        expect(subject).to contain_exactly(course_school_direct_salaried_training_programme)
+      end
+    end
+
+    describe ".with_qualifications" do
+      let(:course_qts) { create(:course, :resulting_in_qts) }
+      let(:course_pgce_with_qts) { create(:course, :resulting_in_pgce_with_qts) }
+      let(:course_pgde_with_qts) { create(:course, :resulting_in_pgde_with_qts) }
+      let(:course_pgce) { create(:course, :resulting_in_pgce) }
+      let(:course_pgde) { create(:course, :resulting_in_pgde) }
```

Comment:
What do you think about adding these, and others, to `TestDataCache`? It'd be nice to start to populate that with commonly-used course types, instead of having to back-fill more and more of these.

Reply:
I've added the `TestDataCache` as discussed offline. Turns out, there were only a small number that worked with it as most in this PR require parameters. I've left it in for now but I'm not sure the overhead is worth it in the unit tests as in this PR. I wonder if we'd be best benchmarking the specs to identify the slowest one and then trying `TestDataCache` on that and getting some data on the performance and seeing the development overhead to add it. We could also try some other approaches on the same file and compare before we spread it too far through the app.